ortb/default_test.go

```diff
@@ -329,3 +337,78 @@ func TestSetDefaultsPriceGranularityRange(t *testing.T) {
 		})
 	}
 }
+
+func TestSetDefaultsImp(t *testing.T) {
+	secure0 := int8(0)
+	secure1 := int8(1)
+
+	testCases := []struct {
+		name             string
+		givenImps        []*openrtb_ext.ImpWrapper
+		expectedImps     []*openrtb_ext.ImpWrapper
+		expectedModified bool
+	}{
+		{
+			name:             "nil",
+			givenImps:        nil,
+			expectedImps:     nil,
+			expectedModified: false,
+		},
+		{
+			name:             "empty",
+			givenImps:        []*openrtb_ext.ImpWrapper{},
+			expectedImps:     []*openrtb_ext.ImpWrapper{},
+			expectedModified: false,
+		},
+		{
+			name:             "one-nil",
+			givenImps:        []*openrtb_ext.ImpWrapper{nil},
+			expectedImps:     []*openrtb_ext.ImpWrapper{nil},
+			expectedModified: false,
+		},
+		{
+			name:             "one-imp-nil",
+			givenImps:        []*openrtb_ext.ImpWrapper{{Imp: nil}},
+			expectedImps:     []*openrtb_ext.ImpWrapper{{Imp: nil}},
+			expectedModified: false,
+		},
+		{
+			name:             "one-imp-secure-0",
+			givenImps:        []*openrtb_ext.ImpWrapper{{Imp: &openrtb2.Imp{Secure: &secure0}}},
+			expectedImps:     []*openrtb_ext.ImpWrapper{{Imp: &openrtb2.Imp{Secure: &secure0}}},
+			expectedModified: false,
+		},
+		{
+			name:             "one-imp-secure-1",
+			givenImps:        []*openrtb_ext.ImpWrapper{{Imp: &openrtb2.Imp{Secure: &secure1}}},
+			expectedImps:     []*openrtb_ext.ImpWrapper{{Imp: &openrtb2.Imp{Secure: &secure1}}},
+			expectedModified: false,
+		},
+		{
+			name:             "one-imp-secure-nil",
+			givenImps:        []*openrtb_ext.ImpWrapper{{Imp: &openrtb2.Imp{Secure: nil}}},
+			expectedImps:     []*openrtb_ext.ImpWrapper{{Imp: &openrtb2.Imp{Secure: &secure1}}},
+			expectedModified: true,
+		},
+		{
+			name:             "one-imp-many-notmodified",
+			givenImps:        []*openrtb_ext.ImpWrapper{{Imp: &openrtb2.Imp{Secure: &secure0}}, {Imp: &openrtb2.Imp{Secure: &secure1}}},
+			expectedImps:     []*openrtb_ext.ImpWrapper{{Imp: &openrtb2.Imp{Secure: &secure0}}, {Imp: &openrtb2.Imp{Secure: &secure1}}},
+			expectedModified: false,
+		},
+		{
+			name:             "one-imp-many-modified",
+			givenImps:        []*openrtb_ext.ImpWrapper{{Imp: &openrtb2.Imp{Secure: &secure0}}, {Imp: &openrtb2.Imp{Secure: nil}}},
+			expectedImps:     []*openrtb_ext.ImpWrapper{{Imp: &openrtb2.Imp{Secure: &secure0}}, {Imp: &openrtb2.Imp{Secure: &secure1}}},
+			expectedModified: true,
+		},
+	}
+
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			actualModified := setDefaultsImp(test.givenImps)
+			assert.Equal(t, test.expectedModified, actualModified)
+			assert.ElementsMatch(t, test.expectedImps, test.givenImps)
```

Comment:
I think I am a little impressed that this passes. The `givenImps` secure element should be pointing to `DefaultSecure`, and not `secure1` as `expectedImps` does, though the values pointed to do match. Something I will have to try to remember about the behavior of `assert.ElementsMatch`.

Reply:
Ah. Surprised by the pointers?

Most of the testify assertions use `reflect.DeepEqual` under the covers which compares values instead of pointer addresses.